src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,344 @@
+    {
+        private readonly int _capacity;
+        // The percentage of the cache to be removed when _maxCapacityPercentage is reached.
+        private readonly double _compactionPercentage = .20;
+        private LinkedList<LRUCacheItem<TKey, TValue>> _doubleLinkedList = new LinkedList<LRUCacheItem<TKey, TValue>>();
+        private BlockingCollection<Action> _eventQueue = new BlockingCollection<Action>();
+        private Task _eventQueueTask;
+        private ConcurrentDictionary<TKey, LRUCacheItem<TKey, TValue>> _map;
+        // When the current cache size gets to this percentage of _capacity, _compactionPercentage% of the cache will be removed.
+        private readonly double _maxCapacityPercentage = .95;
+        private bool _disposed = false;
+        private readonly int _tryTakeTimeout;
+        // if true, expired values will not be added to the cache and clean-up of expired values will occur on a 5 minute interval
+        private readonly bool _removeExpiredValues;
+
+        internal EventBasedLRUCache(
+            int capacity,
+            TaskCreationOptions options = TaskCreationOptions.LongRunning,
+            IEqualityComparer<TKey> comparer = null,
+            int tryTakeTimeout = 500,
+            bool removeExpiredValues = false,
+            int cleanUpIntervalInSeconds = 300)
+        {
+            _tryTakeTimeout = tryTakeTimeout;
+            _capacity = capacity > 0 ? capacity : throw LogHelper.LogExceptionMessage(new ArgumentOutOfRangeException(nameof(capacity)));
+            _map = new ConcurrentDictionary<TKey, LRUCacheItem<TKey, TValue>>(comparer ?? EqualityComparer<TKey>.Default);
+            _removeExpiredValues = removeExpiredValues;
+            _eventQueueTask = new Task(OnStart, options);
+            _eventQueueTask.Start();
+            if (_removeExpiredValues)
+                _ = RemoveExpiredValuesPeriodically(TimeSpan.FromSeconds(cleanUpIntervalInSeconds));
+        }
+
+        private void OnStart()
+        {
+            while (!_disposed)
+            {
+                try
+                {
+                    if (_eventQueue.TryTake(out var action, _tryTakeTimeout))
+                        action.Invoke();
+                }
+                catch (Exception ex)
+                {
+                    LogHelper.LogWarning(LogHelper.FormatInvariant(LogMessages.IDX10900, ex));
+                }          
+            }
+        }
+
+        public bool Contains(TKey key)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            return _map.ContainsKey(key);
+        }
+
+        /// <summary>
+        /// FOR TESTING PURPOSES ONLY.
+        /// </summary>
+        internal void WaitForProcessing()
+        {
+            while (true)
+            {
+                if (_eventQueue.Count == 0)
+                    return;
+            }
+        }
+
+        internal int RemoveExpiredValues()
```

Comment:
>RemoveExpiredValues [](start = 21, length = 19)

perhaps we should check for disposed and also have a try catch around this entire operation, during dispose this could fail?